#Listing libraries for loading packages
library(DBI)
library(RODBC)
library(odbc)
library(dplyr)
library(dbplyr)
library(RPostgreSQL)
library(shiny)
library(shinythemes)
library(markdown)
library(DT)
library(tidyverse)
#Registering a user interface with Shiny
shinyUI(fluidPage(theme = shinytheme("yeti"),
  #Creates the tabset panel system for organizing home, about, gene search, 
  #analyze data, references, and help page
  tabsetPanel(id = "msitabs",
    #Home page welcomes user to the web app and provides a link button to the 
    #gene search page
    tabPanel("Home",
      sidebarLayout(
        sidebarPanel(width = 6,
          h4("Welcome to the MSI Gene Database"), 
          h5("Web Development in Advanced Bioinformatics"),
          actionButton("homesearch", "Go To Search")),
        mainPanel(width = 6
        )
      )
    ),
    #About page describes the backend database, scoring algorithm, 
    #as well as frontend search page for the MSI gene database webapp
    tabPanel("About",
      sidebarLayout(
        sidebarPanel(width = 6,
          h4("About MSI Gene Database"),
          h5("The MSI Gene Database is a well-organized webapp tool to visualize and analyze gene expression data related to microsatellite instability in regions where frameshift mutations generate shifted peptides which are used as variable neoantigen markers for cancer cells. Predictions are made for clustering groups of neoantigens based on their variations in chromosomal DNA."),
          h5("The webapp tool integrates with a PostgreSQL object-relational database created with a relational model that contains specific genes and their MSI values. There is a scoring system based on the level of mutation that the neoantigen has from the original FSP gene sequence."),
          h5("Predicted neoantigens are displayed alongside their clustering value and range for a specific FSP derived from the MSI bait gene sequence. One of the tables created is for storing a list of UP-regulated MSI genes, and their gene expression value obtained from the gene expression table of the GEO microarray experiment. An additional table stores the MSI bait gene-set sequences of the FSPs alongside the name of the MSI gene that the sequences belong to. The second relational table contains names of significant background genes obtained from the gene expression table for the GEO microarray experiment alongside GEO provided MSI values for those gene names."),
          h5("The web application takes user input and validates it against a database of 100 top FSP replacements and associated properties for a list of 13 oncogenes. MSI scores are generated by an algorithm programmed in a Jupyter notebook and stored in the database alongside properties of the FSP found in the samples studied. The algorithm locates FSP neoantigens by analyzing sequences for CpG islands and mononucleotide sequences approximately 20 base pairs upstream of the coding sequence CDS. The algorithm is functional for a sample size of approximately 100 with a runtime of approximately 16 hours. However, for the sake of brevity, a sample size threshold was set for this experiment at 5 samples that are known to contain FSPs."),
          h5("This webapp will display and graph selected FSP neoantigens from target MSI genes."),
          h5("The data is searchable by gene name and MSI-categorization with the ability to limit search results."),
          h5("RShiny is Functional")
        ),
        mainPanel(width = 6
        )
      )
    ),
    #Allows the user to search the MSI gene database by entering a MSI gene name
    #Results can then be filtered by MSI-categorization (high, low, and ambiguous)
    #A 1, 5, 10, 50, 100 result limitation can be placed on the search
    tabPanel("Gene Search", 
      sidebarLayout(
        sidebarPanel(width = 6,       
          h3("Gene Search:"), h3("MSI Gene Database Search Form"), 
          h5("MSI Categorization"), selectInput("var1", "", choices = c("MSI-high", "MSI-low", "MSI-ambiguous")), 
          h5("Enter MSI Gene Name"), textInput("name", "", ""),
          h5("Limit Search Results By"), selectInput("var2", "", choices = c("1", "5", "10", "50", "100")), 
          actionButton("Genesearch", "Submit")),
        mainPanel(width = 6,
          h3("MSI Gene Database Search Results"),
          h4("List of MSI Genes"),
          DT::dataTableOutput("table")
        )
      )
    ),
    #This page allows users to enter the gene name of the MSI gene they want to
    #analyze and select the MSI-categorization that they want to filter by (high,
    #low, ambiguous)
    tabPanel("Analyze Data",
      sidebarLayout(
        sidebarPanel(width = 12,
          h3("Graphic Analysis:"), 
          h3("MSI Gene Graphic Analysis Form"), h5("MSI Categorization"), 
          selectInput("var", "", choices = c("MSI-high", "MSI-low", "MSI-ambiguous")),
          h5("Enter MSI Gene Name"), textInput("dataname", "", ""),
          actionButton("Analyzedata", "Analyze Data")),
        mainPanel(width = 0)
      )
    ),
    #Displays two figures with descriptions below the figure
    #Figure one is a histogram that plots MSI values for select neoantigens 
    #against the frequency of occurrence
    #Figure two is a pie chart that plots the number of neoantigens retrieved 
    #from each sample
    tabPanel("Analyze Data Figures",
      sidebarLayout(
        sidebarPanel(width = 0),
        mainPanel(width = 12,
          uiOutput('output1'),
          plotOutput("myhist"),
          uiOutput('output2'),
          uiOutput('output3'),
          uiOutput('output4'),
          br(),
          plotOutput("mypi"),
          uiOutput('output5'),
          uiOutput('output6'),
          uiOutput('output7'),
          br(),
          fluidRow(
            column(6, align="center", offset=3,
              actionButton("backsearch", "Back To Search")),
          ),
          br()
        )
      )
    ),
    #Displays the research articles and their associated links along with 
    #software libraries, programming tools/languages etc. used for retrieving
    #data and building the webapp that accesses the MSI gene database
    tabPanel("References",
      sidebarLayout(
        sidebarPanel(width = 7,
          h4("References for MSI Gene Database"), 
          h5("Research Articles"),
          h5(HTML("<p>Ballhausen, Alexej, et al. “The Shared Frameshift Mutation Landscape of Microsatellite-</p> 
          <p>&emsp;Unstable Cancers Suggests Immunoediting during Tumor Evolution.” <i>Nature</i></p> 
          <p>&emsp;<i>Communications</i>, U.S. National Library of Medicine, 21 Sept. 2020,</p>")),
          h5(tags$a(href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7506541/", target = "_blank", HTML("<p>&emsp;https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7506541/.</p>"))),
          h5(HTML("<p>Gebert J;Gelincik O;Oezcan-Wahlbrink M;Marshall JD;Hernandez-Sanchez A;Urban K;Long</p> 
          <p>&emsp;M;Cortes E;Tosti E;Katzenmaier EM;Song Y;Elsaadi A;Deng N;Vilar E;Fuchs V;Nelius</p> 
          <p>&emsp;N;Yuan YP;Ahadova A;Sei S;Shoemaker RH;Umar A;Wei L;Liu S;Bork P;Edelmann</p>
          <p>&emsp;W;von Knebel Doe. “Recurrent Frameshift Neoantigen Vaccine Elicits Protective</p>
          <p>&emsp;Immunity with Reduced Tumor Burden and Improved Overall Survival in a Lynch</p>
          <p>&emsp;Syndrome Mouse Model.” <i>Gastroenterology</i>, U.S. National Library of Medicine,</p>")),
          h5(tags$a(href="https://pubmed.ncbi.nlm.nih.gov/34224739/", target = "_blank", HTML("<p>&emsp;https://pubmed.ncbi.nlm.nih.gov/34224739/.</p>"))),
          h5(HTML("<p>“Geo Accession Viewer.” <i>National Center for Biotechnology Information</i>, U.S. National Library</p> 
          <p>&emsp;of Medicine, <a href='https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE51334', target = '_blank'>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE51334.</a></p>")),
          h5(HTML("<p>“Geo Accession Viewer.” <i>National Center for Biotechnology Information</i>, U.S. National Library</p> 
          <p>&emsp;of Medicine, <a href='https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE162960', target = '_blank'>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE162960.</a></p>")),
          h5(HTML("<p>Rathe SK;Johnson JE;Silverstein KA;Erdmann JJ;Watson AL;Popescu FE;Ohlfest</p> 
          <p>&emsp;JR;Largaespada DA; “MMUFLR: Missense Mutation and Frameshift Location Reporter.”</p> 
          <p>&emsp;<i>Bioinformatics (Oxford, England)</i>, U.S. National Library of Medicine,</p>")),
          h5(tags$a(href="https://pubmed.ncbi.nlm.nih.gov/23825368/", target = "_blank", HTML("<p>&emsp;https://pubmed.ncbi.nlm.nih.gov/23825368/.</p>"))),
          h5(HTML("<p>Suvorova, Y M. “Search for Potential Reading Frameshifts in Cds from Arabidopsis Thaliana</p> 
          <p>&emsp;and Other Genomes.” <i>Academic.oup.com</i>,</p>")),
          h5(tags$a(href="https://academic.oup.com/dnaresearch/article/26/2/157/5306610", target = "_blank", HTML("<p>&emsp;https://academic.oup.com/dnaresearch/article/26/2/157/5306610.</p>"))),
          h5(HTML("<p>Veigl, Martina L. et. Al. “Biallelic Inactivation of hMLH1 by Epigenetic Gene Silencing, a Novel</p>
          <p>&emsp;Mechanism Causing Human MSI Cancers.” <i>The Proceedings of the National Academy of</p>
          <p>&emsp;Sciences</i>, 21 July 1998, <a href='https://www.pnas.org/doi/10.1073/pnas.95.15.8698', target = '_blank'>https://www.pnas.org/doi/10.1073/pnas.95.15.8698.</a></p>")),
          h5("Data Sources"),
          h5(HTML("<ul> <li>GEO (Gene Expression Omnibus)</li> <li>NCBI (National Center for Biotechnology Information) – Gene</li></ul>")),
          h5("Software Libraries"),
          h5(HTML("<ul> 
                  <li>Python Bio.Entrez</li> 
                  <li>Python Bio.Align</li> 
                  <li>Python CSV</li> 
                  <li>R DBI</li>
                  <li>R RODBC</li>
                  <li>R odbc</li>
                  <li>R dplyr</li>
                  <li>R dbplyr</li>
                  <li>R RPostgreSQL</li>
                  <li>R shiny</li>
                  <li>R shinythemes</li>
                  <li>R markdown</li>
                  <li>R DT</li>
                  <li>R tidyverse</li>
                  <li>R seqinr</li>
                  <li>R Biostrings</li>
                  </ul>")),
          h5("Programming Tools/Languages etc."),
          h5(HTML("<ul> 
                  <li>R</li> 
                  <li>Python</li> 
                  <li>HTML (RShiny)</li> 
                  <li>CSS (RShiny)</li>
                  <li>Javascript (RShinyjs)</li>
                  </ul>")),
          h5(HTML("<p>&emsp;<u>Source Code Editor:</u> Atom Text Editor, Jupyter Notebook, RStudio</p>")),
          h5(HTML("<p>Final Project GIT Repository: <a href='https://github.com/jfosterjr99/Bioinformatics-Senior-Capstone-Project', target = '_blank'>https://github.com/jfosterjr99/Bioinformatics-Senior-Capstone-Project</a></p>"))
        ),
        mainPanel(width = 5
        )
      )
    ),
    #Contains help for executing searches in the gene search and analyze data pages
    #Contains a list of MSI genes that are searchable in database
    #Explains the system for categorizing neoantigens by high, ambiguous, 
    #and low scores
    tabPanel("Help",
      sidebarLayout(
        sidebarPanel(width = 6,
          h4("Help for MSI Gene Database"),
          h5("This page contains help for executing searches and analyzing results"),
          h5("The data is searchable by MSI categorization and gene name with the ability to limit search results"),
          tableOutput("table2"),
          h6(HTML("<b>Genes affected by cMNR FS mutations and derived FSPs. 
             Predicted immunoscores for neoantigens aligned with a 7-10 
             amino acid wild-type (WT) bait sequence for genes on C57BL/6 alleles 
             H2-Db and H2-Kb (with assosciated SYFPEITHI scoring). The numbers of MSI-high, MSI-low, and MSI-ambiguous are calculated from low numbers of percentage rank (<1), as well as MSI values > SYFPEITHI scores, predict MSI-high neoantigens.</b>"))
        ),
        mainPanel(width = 6
        )
      )
    )
  )
)
)