{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f909179d-3892-483c-a753-7315892e83db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n",
      "Done.\n",
      "Done.\n",
      "Done.\n",
      "Done.\n",
      "nacad done.\n",
      "xirp1 done.\n",
      "5730596b20rik done.\n",
      "rif1 done.\n",
      "maz done.\n",
      "hic1 done.\n",
      "sdccag1 done.\n",
      "tmem107 done.\n",
      "srcin1 done.\n",
      "marcks done.\n",
      "senp6 done.\n",
      "phactr4 done.\n",
      "chrnb2 done.\n"
     ]
    }
   ],
   "source": [
    "#Importing align package from biopython\n",
    "from Bio import Align\n",
    "from Bio.Align import substitution_matrices\n",
    "\n",
    "#Creating a dictionary with MSI gene names as keys and lists to store tuples of alignment scores, alignments, SRR file names, \n",
    "#SRR positions, CpG island counts, and mononucleotide sequences\n",
    "mylist = {'nacad': [], 'xirp1': [], '5730596b20rik': [], 'rif1': [], 'maz': [],\n",
    "          'hic1': [], 'sdccag1': [], 'tmem107': [], 'srcin1': [], 'marcks': [],\n",
    "          'senp6': [], 'phactr4': [], 'chrnb2': []}\n",
    "#Creating a list of SRR file names that will be searched for FSPs\n",
    "mySRRList = ['SRR14676234', 'SRR14676248', 'SRR14676209', 'SRR14676235', 'SRR14676227']\n",
    "for SRRfile in mySRRList:\n",
    "    #Reading in the files that contain the 7 amino acid long peptide sequences alongside the file that contains the \n",
    "    #corresponding chromsome positions, CpG island counts, and mononucleotide sequences for those specific peptide sequences\n",
    "    #with a count variable to establish which chromsome positions, CpG island counts, and mononucleotide sequences match the\n",
    "    #the specific peptide sequence being observed in the file\n",
    "    f = open(SRRfile + '/' + SRRfile + 'format3.txt', 'r')\n",
    "    f2 = open(SRRfile + '/' + SRRfile + 'format2.txt', 'r')\n",
    "    lines = f2.readlines()\n",
    "    count = 0\n",
    "    for row in f:\n",
    "        #A dictionary to contain the FSP bait sequences for the 13 MSI genes used in the study\n",
    "        fspSequences = {'nacad': 'VIYAPPP', 'xirp1': 'GKGPGGPP', '5730596b20rik': \n",
    "                       'GTLPPPP', 'rif1': 'AHTDKKKK', 'maz': 'PCTLLAP', 'hic1': 'DRTFPSPP',\n",
    "                       'sdccag1': 'EAPKGKKK', 'tmem107': 'TQYFGMGG', 'srcin1': 'DEGMWPP', 'marcks': 'SSETPKKK',\n",
    "                       'senp6': 'VKCSMKKK', 'phactr4': 'PWKWRKKK', 'chrnb2': 'VRTRPSPP'}\n",
    "        #Generating alignments using a BLOSUM62 scoring matrix for each 7 amino acid long peptide sequence crossed with each \n",
    "        #FSP bait sequence in the dictionary containing the 13 MSI genes\n",
    "        for key in fspSequences:\n",
    "            aligner = Align.PairwiseAligner()\n",
    "            aligner.substitution_matrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "            alignments = aligner.align(str(fspSequences[key]), row[1:len(row)-2])\n",
    "            alignments = list(alignments)\n",
    "            alignment = alignments[0]\n",
    "            \n",
    "            line = lines[count].split('\\t')\n",
    "            srrposition = line[0]\n",
    "            cpgcount = line[1]\n",
    "            mononucsequence = line[2]\n",
    "            mylist[key].append((alignment.score, alignment, SRRfile, srrposition, cpgcount, mononucsequence))\n",
    "        count+=1\n",
    "    f2.close()\n",
    "    f.close()\n",
    "    print(\"Done.\")\n",
    "\n",
    "#Find the 100 highest scoring alignments for each MSI gene in the 'mylist' dictionary\n",
    "for key in mylist:\n",
    "    while(len(mylist[key]) > 100):\n",
    "        min = float('inf')\n",
    "        for entry in mylist[key]:\n",
    "            if(entry[0] < min):\n",
    "                min = entry[0]\n",
    "        for entry in mylist[key]:\n",
    "            if(entry[0] == min):\n",
    "                mylist[key].remove(entry)\n",
    "                break\n",
    "    print(key, \"done.\")\n",
    "#Generating a new file to store each possible neoantigen replacement found denoted by its MSI gene name, MSI value generated from \n",
    "#the number of CpG islands with a 1000x multiplier and the length of the longest mononucleotide sequence found 20 nucleotide bases \n",
    "#upstream of the CDS with a multiplier of 100x, the upper alignment, aswell as the lower alignment, SRR file name, chromosome position, \n",
    "#and mononucleotide sequence\n",
    "f1 = open('fileOutScore.txt', 'a')\n",
    "for key in mylist:\n",
    "    for entry in mylist[key]:\n",
    "        MSI_Val = entry[0] + (int(entry[4])*1000) + (len(entry[5])*100)\n",
    "        f1.write(str(key) + '\\t' + str(MSI_Val) + '\\t' + str(entry[1]).split('\\n')[0] + '\\t' + str(entry[1]).split('\\n')[2] + '\\t' + str(entry[2]) + '\\t' + str(entry[3]) + '\\t' + str(entry[5]))\n",
    "f1.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2aff10a5-01ef-466a-96de-a88d519230a9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
